### PR TITLE
Terratest test framework

### PR DESCRIPTION
**Pull request by vara-bonthu**
### What does this PR do?

- Terratest test framework for one example
- License file updated for AWS-IA

### More

- [x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

**Terratest Test Results** 

```TestEksBlueprintsE2E 2022-04-12T11:28:38+01:00 logger.go:66: Destroy complete! Resources: 0 destroyed.
--- PASS: TestEksBlueprintsE2E (0.00s)
    --- PASS: TestEksBlueprintsE2E/TF_PLAN_VALIDATION (43.59s)
    --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION (17.48s)
        --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION/vpc_cidr (3.53s)
        --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION/vpc_private_subnet_cidr (3.40s)
        --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION/vpc_public_subnet_cidr (3.40s)
        --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION/eks_cluster_id (3.67s)
        --- PASS: TestEksBlueprintsE2E/TF_OUTPUTS_VALIDATION/eks_managed_nodegroup_status (3.47s)
    --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION (8.90s)
        --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/MATCH_EKS_CLUSTER_NAME (0.00s)
        --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/MATCH_TOTAL_EKS_WORKER_NODES (0.00s)
        --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION (0.24s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/MATCH_REPLICAS_VS_READY-REPLICAS/aws-load-balancer-controller (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/UNAVAILABLE_REPLICAS/aws-load-balancer-controller (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/MATCH_REPLICAS_VS_READY-REPLICAS/cluster-autoscaler-aws-cluster-autoscaler (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/UNAVAILABLE_REPLICAS/cluster-autoscaler-aws-cluster-autoscaler (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/MATCH_REPLICAS_VS_READY-REPLICAS/coredns (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/UNAVAILABLE_REPLICAS/coredns (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/MATCH_REPLICAS_VS_READY-REPLICAS/metrics-server (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DEPLOYMENTS_VALIDATION/UNAVAILABLE_REPLICAS/metrics-server (0.00s)
        --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DAEMONSETS_VALIDATION (0.12s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DAEMONSETS_VALIDATION/MATCH_DESIRED_VS_CURRENT_PODS/aws-node (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DAEMONSETS_VALIDATION/UNAVAILABLE_REPLICAS/aws-node (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DAEMONSETS_VALIDATION/MATCH_DESIRED_VS_CURRENT_PODS/kube-proxy (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_DAEMONSETS_VALIDATION/UNAVAILABLE_REPLICAS/kube-proxy (0.00s)
        --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_SERVICES_VALIDATION (0.24s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_SERVICES_VALIDATION/SERVICE_STATUS/cluster-autoscaler-aws-cluster-autoscaler (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_SERVICES_VALIDATION/SERVICE_STATUS/kube-dns (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_SERVICES_VALIDATION/SERVICE_STATUS/kubernetes (0.00s)
            --- PASS: TestEksBlueprintsE2E/EKS_ADDON_VALIDATION/EKS_SERVICES_VALIDATION/SERVICE_STATUS/metrics-server (0.00s)
    --- PASS: TestEksBlueprintsE2E/eks-cluster-with-new-vpc (1644.63s)
PASS
ok      github.com/aws-ia/terraform-aws-eks-blueprints/src      1645.261s
```

### CODE REVIEW

File: data.tf [hunk 1]
```diff
@@ -9,15 +9,6 @@ data "aws_eks_cluster" "cluster" {
...
 
-data "aws_eks_cluster_auth" "cluster" {
-  count = var.create_eks ? 1 : 0
-  name  = module.aws_eks.cluster_id
-}
-
-data "aws_availability_zones" "available" {
-  state = "available"
-}
-
```
Comment (vara-bonthu) [inline data.tf:20]: Not being used anywhere in the code

File: locals.tf [hunk 2]
```diff
@@ -23,10 +23,8 @@ locals {
   public_subnet_ids  = var.public_subnet_ids
   tags               = module.eks_tags.tags
 
-  enable_workers                    = length(var.self_managed_node_groups) > 0 || length(var.managed_node_groups) > 0 ? true : false
-  cluster_security_group_id         = module.aws_eks.cluster_security_group_id
-  cluster_primary_security_group_id = module.aws_eks.cluster_primary_security_group_id
```
Comment (vara-bonthu) [inline locals.tf:28]: Not used anywhere in the code

File: test/src/eks_api.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+func newEksSession(region string) *eks.EKS {
+	mySession := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(region),
+	}))
+	svc := eks.New(mySession)
+	return svc
+}
+
+func GetKubernetesClient(cluster *eks.Cluster) (*kubernetes.Clientset, error) {
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply test/src/eks_api.go]: Only `internal methods` in Go uses small case in Function Names to start with. All the external methods(imported through packages) uses the Uppercase to start with.

I have optimised further to make `eks.go` as an internal package for better readability of the code.

File: .github/workflows/e2e-terratest.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+name: e2e-test
```
Comment (kcoleman731) [inline .github/workflows/e2e-terratest.yml:1]: change name to e2e-terratest

File: test/src/eks_blueprints_e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+	"testing"
+	internal "github.com/aws-ia/terraform-aws-eks-blueprints/aws"
+)
+
+var (
+	//Test Driven tests Inputs https://github.com/golang/go/wiki/TableDrivenTests
+	testCases = []struct {
```
Comment (kcoleman731) [inline test/src/eks_blueprints_e2e_test.go:21]: should we type this struct?

File: test/src/eks_blueprints_e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+	S3BackendConfig = map[string]string{
+		"bucketName": "terraform-ssp-github-actions-state",
+		"s3Prefix": "terratest/examples/",
+		"awsRegion" : "us-west-2"}*/
+
+	destroyModules = []string{
+		"module.kubernetes_addons",
+		"module.eks_blueprints",
+		"module.aws_vpc",
+		"full_destroy"}
```
Comment (kcoleman731) [inline test/src/eks_blueprints_e2e_test.go:46]: closing bracket on new line

**Review by kcoleman731 — COMMENTED** [review]
Super cool @vara-bonthu . Small comments

**Review by askulkarni2 — COMMENTED** [review]
LGTM! Please address @kcoleman731's comments, then we are good to merge.

File: test/src/eks_blueprints_e2e_test.go (hunk 5 above)
Comment (vara-bonthu) [reply test/src/eks_blueprints_e2e_test.go:21]: Just using the standard table driven struct from Go tests. I will explore this bit more in my next improvement PR

File: .github/workflows/e2e-terratest.yml (hunk 4 above)
Comment (vara-bonthu) [reply .github/workflows/e2e-terratest.yml]: done

File: LICENSE [hunk 7]
```diff
@@ -1,14 +1,201 @@
...
-use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
-the Software, and to permit persons to whom the Software is furnished to do so.
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
-FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
-COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
-IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
-CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+                                 Apache License
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply LICENSE]: done

File: test/src/eks_blueprints_e2e_test.go (hunk 6 above)
Comment (vara-bonthu) [reply test/src/eks_blueprints_e2e_test.go]: Thanks for spotting this. `Go fmt`  magically fixed the issue

**Review by askulkarni2 — APPROVED** [review]
🚢

**Pull request merged**